[Backend] mongo connector working now

markwitz opened the pull request:
resolve #87 
I had to create a separate container starting the replica set dunno if there is a better way to avoid this but it's working like this for me. Es is getting updated when I insert to mongo.
Still needs testing:
- [x] the id issue how is it handled  by the connector ?
- [x] can we find projects by ID?
- [x]  at what rate are we syncing ? and can we configure it?
- [x] what about including excluding collections? account data?

Commit:
added new mongo-connector container, can't connect yet though

Commit:
updated addresses in the Dockerfile to be the correct ones

Commit:
updated config file to actually stream the warnings and errors

Commit:
tried replicating mongodb with only one replica, not working so far

Commit:
tried replicating mongodb with only one replica, not working so far

Commit:
new try to connect mongo-connector

Commit:
small edit to mongo-connector-config

markwitz committed:
made connector working

 - mongodb is now created as an replica set
 - mongo connector ist now started after the replica set is created and sleeping for 15
 - mongodb/setup.sh is executed in a seperate docker container after mongodb is started

henderkes commented on backend/mongo_connector/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM python:3.4
```
better remove the subversion specification

markwitz replied:
its stable for 3.4 so i specified it

henderkes reviewed (commented):
the whole thing with the replica set being required is normal, I will have a deeper look at this on the weekend but it looks very good for now.

could anyone working on linux/docker-toolbox try to pull the branch, set it up locally and post back results?

markwitz committed:
lowerd verbosity to make congo connector more readable

markwitz committed:
port passing in docker compose for better testing

markwitz committed:
added test for database consistency and post consistency

this checks if th data is the same for  es and mongo

markwitz committed:
longer sleep for travis

markwitz committed:
mongo_client is now getting the db correctly

markwitz committed:
slave ok added to setup

markwitz committed:
try some travis debugging with script that waits for mongo

markwitz commented:
I don't really know whats happening but the test that has passed at first will no longer run on travis even those without changes. I cant recreate locally

https://travis-ci.org/Drakulix/knex/jobs/242851069 passed but when re reran for the same commit it would not pass and throw me some error

markwitz committed:
added another ps for better debugging

markwitz committed:
sleep after connected so replica can be enabled

markwitz committed:
enables build in copose up

markwitz committed:
make setup replica allways restart

henderkes commented:
Apparently "all" tests fail so there's most likely some internal server error going on (with the connector right after storing a manifest in mongodb?). We simply need better debugging tools than pytest, flask prints and travis output.

markwitz commented:
the container that should initialise the repSet is constantly restarting but i cant really tell why because i can't reproduce the error on my machine

henderkes commented:
Strange enough this works wonderfully for me locally, too

henderkes committed:
Merge branch 'master' into Backend/feature/mongo-connector/iss87

henderkes committed:
changed gitignore

henderkes committed:
added ''

henderkes committed:
fixed inconsistent test

henderkes committed:
attempt at only restarting replica set on failure

henderkes committed:
dont restart replica on failure

henderkes committed:
debugging ps

henderkes committed:
fix pep8

henderkes committed:
removed replset to test if tests can connect to mongo again

henderkes committed:
28 passed in 11.62 seconds

henderkes commented:
![image](https://user-images.githubusercontent.com/7896469/27252850-72a2e506-5368-11e7-84d9-dae1d57473b5.png)

hmm..

henderkes committed:
reset

henderkes commented:
Okay, so mongodb service fails and cannot be connected to if the --replSet option is specified. Therefore the replicaset image won't run and the connector won't work. The question is why the --replSet option makes mongodb fail in travis but not locally. Please someone with linux test this,

henderkes committed:
Merge branch 'master' into Backend/feature/mongo-connector/iss87

henderkes commented:
We can set the '--auto-commit-interval=0' flag to make every write operation in mongodb to be instantly synced to elasticsearch.

The mongodb['_id'] field becomes the elasticsearch id field.
excluding collections is done in the connector config in the backend/mongo_connector folder. We should probably exclude the user stuff, but I'm not sure in which collection it is?

henderkes committed:
Merge branch 'master' into Backend/feature/mongo-connector/iss87

henderkes committed:
excluded user collection

henderkes committed:
Merge branch 'Backend/feature/mongo-connector/iss87' of https://github.com/Drakulix/knex into Backend/feature/mongo-connector/iss87

henderkes commented:
Excluded the user collection, everything works flawlessly now except for travis...

Drakulix committed:
Fixup docker replica

Drakulix committed:
Add ports for tests again

henderkes committed:
knexDB to knexdb

henderkes committed:
Merge branch 'Backend/feature/mongo-connector/iss87' of https://github.com/Drakulix/knex into Backend/feature/mongo-connector/iss87

henderkes committed:
404 to 405

henderkes commented:
From Mongoconnector docs: 

Elasticsearch Indexes, Mappings, and Types

> Mongo Connector gives each MongoDB database its own index in Elasticsearch. Each MongoDB collection becomes its own mapping type. For example, documents from the collection kittens in the database animals will put into the animals index in Elasticsearch with a mapping type of kittens

the index knexdb is not generated, therefore the connector is not working yet.

Drakulix committed:
Index naming fixed

henderkes committed:
Merge branch 'master' into Backend/feature/mongo-connector/iss87

Drakulix committed:
Fix travis.yml

Drakulix committed:
Fix travis wait script

Drakulix committed:
Change index for flask-security as well

Drakulix committed:
Add small sleep after mongo startup again

henderkes committed:
something in the dockerfile

henderkes committed:
fixed replica setup

henderkes committed:
hopefully correct now

henderkes committed:
fixed false assumption that we ran a secondary repset

henderkes committed:
removed last 3 lines in setupsh

Drakulix committed:
correctly init replSet

Drakulix committed:
Cleanup

- Remove unnecessary links flags
- Use not deprecated elastic search image
- Set elastic search config parameters
- Let mongo-connector sleep for correct startup

Drakulix committed:
container startup time

Drakulix committed:
moar output

Drakulix committed:
set node priorities

Drakulix committed:
Remove unnecessary python versions

henderkes committed:
rewrite delete call

henderkes committed:
removed python 3.3-3.5 for now

Drakulix committed:
More fixes

Use docker mongo connector, that waits correctly
Wait for the right mongo container to do the replica setup
Adjust config for namespaces
Add docker dependencies

Drakulix committed:
Readd auto interval

henderkes commented on backend/mongo_connector/mongo-connector-config.json:
```diff
@@ -12,7 +12,8 @@
         "docManager" : "elastic2_doc_manager",
-        "targetURL" : "elasticsearch:9200"
+        "targetURL" : "elasticsearch:9200",
+	"autoCommitInterval" : 0
```
missing a space

Drakulix committed:
moar bugfix

Drakulix committed:
Remove useless script

Drakulix committed:
Try adding whitelist

Drakulix committed:
Merge branch 'master' into Backend/feature/mongo-connector/iss87

henderkes commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+FROM ubuntu:latest
```
we should really replace this by something more lightweight in the future, but it does the job for now

henderkes commented on backend/flask_api.py:
```diff
@@ -31,34 +31,33 @@
-CORS(app)
 
-with app.open_resource("manifest_schema.json") as schema_file:
+with open("manifest_schema.json") as schema_file:
```
we should revert this to app.open_resource() and see what the problem with str/bytes was

henderkes commented on backend/flask_api.py:
```diff
@@ -31,34 +31,33 @@
-app.config['UPLOAD_FOLDER'] = ''
-app.config['MAX_CONTENT_PATH'] = 1000000  # 100.000 byte = 100kb
-
 
```
One blank line too much, but really not important

henderkes commented on backend/flask_api.py:
```diff
@@ -31,34 +31,33 @@
 # MongoDB Config
-app.config['MONGODB_DB'] = 'knexDB'
+app.config['DEBUG'] = True
+app.config['SECRET_KEY'] = 'super-secret'
```
should probably replace this is something more professional

henderkes commented on backend/flask_api.py:
```diff
@@ -31,34 +31,33 @@
 app.config['MONGODB_HOST'] = 'mongodb'
 app.config['MONGODB_PORT'] = 27017
 app.config['SECURITY_PASSWORD_HASH'] = 'pbkdf2_sha512'
 app.config['SECURITY_PASSWORD_SALT'] = 'THISISMYOWNSALT'
```
see above

henderkes commented on backend/flask_api.py:
```diff
@@ -335,8 +329,7 @@ def search():
     try:
-        res = es.search(index="projects-index",
-                        doc_type="Project", body=request.json)
+        res = es.search(index="knexdb", doc_type="Project", body=request.json)
```
we should remove doc_type specification and this should be request.get_json()

henderkes commented on backend/mongodb_replica_setup/setup.sh:
```diff
@@ -0,0 +1,21 @@
+            {
+                "_id": 0,
+                "host": "mongodb:27017",
+		"priority": 1,
```
missing one space here?

henderkes commented on docker-compose.yml:
```diff
@@ -3,17 +3,38 @@ version: '3'
 
     elasticsearch:
-        image: elasticsearch
+        image: docker.elastic.co/elasticsearch/elasticsearch:5.4.1
```
reason for this change?

henderkes commented on docker-compose.yml:
```diff
@@ -3,17 +3,38 @@ version: '3'
+            - "action.auto_create_index=true"
         ports:
             - "9200:9200"
+        restart: "on-failure"
```
we should test whether an elasticsearch error inducing a shutdown returns a failure code, mongo-connector for example doesn't and has to be on "always"

henderkes commented on tests/api/test_database_connection.py:
```diff
@@ -18,3 +20,23 @@ def test_empty_database(flask_api_url):
+    # Test if document is in mongo
+    assert mongo_result == dummy_project
+    # wait for the refreshed elastic database
+    time.sleep(30)
```
sleep(30) is a little long, 5 seconds should do

henderkes commented on tests/api/test_projects.py:
```diff
@@ -220,16 +252,26 @@ def test_success(self, flask_api_url, pytestconfig):
         print(post_response.text)
         for id in post_response.json():
             assert UUID(id, version=4)
+        project_id = post_response.json()[0]
```
this line is unnecessary, we should instead assert post_response.status_code == 200

henderkes reviewed (commented):
overall very well, I will adress most changes myself, but would like input on the docker-compose config one

henderkes committed:
adressed review changes

henderkes committed:
fix date_creation override and es creates in uploader

henderkes committed:
turn off notifications..

henderkes committed:
add json5s from sc, set date_creation if not specified

henderkes committed:
fixing variable initialization order, importing required mongoengine fields

henderkes committed:
remove first initialisation of "db"

henderkes committed:
fixed pose/post typo

markwitz commented on backend/flask_api.py:
```diff
@@ -22,43 +22,40 @@
-coll = db.projects
-app = Flask(__name__)
+# Create app
+app = Flask(__name__, static_url_path='')
```
what does static_url_path do?

henderkes replied:
no idea, this isn't from me, from docs:

> static_url_path – can be used to specify a different path for the static files on the web. Defaults to the name of the static_folder folder.

> static_folder – the folder with static files that should be served at static_url_path. Defaults to the 'static' folder in the root path of the application.

Drakulix replied:
If of interested, this basically means, that static files should be served from the root of the webside, instead of "/static/" as is default.

markwitz commented on backend/flask_api.py:
```diff
@@ -22,43 +22,40 @@
 app.config['SECRET_KEY'] = 'super-secret'
-
-login_manager = LoginManager()
-login_manager.init_app(app)
```
why are we dropping this?

henderkes replied:
it's not dropped, it's relocated

markwitz replied:
ok where is it now?

markwitz commented on backend/flask_api.py:
```diff
@@ -82,7 +79,7 @@ class User(db.Document, UserMixin):
 class EmailConverter(BaseConverter):
     regex = r"([a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\." +\
-            r"[a-z0-9!#$%&'*+\/=?^_`""{|}~-]+)" +\
+            r"[a-z0-9!#$%&'*+\/=?^_`"r"{|}~-]+)" +\
```
why are we adding the "r" here?

henderkes replied:
make it a raw string literal, it isn't necessary there but goes in style with the other r's I added (which I had to, \ and other letters have to be escaped otherwise which they weren't)

markwitz reviewed (commented):
I had only a few questions

henderkes reviewed (approved):
GREEN
